service/src/main/java/org/databiosphere/workspacedataservice/dataimport/PfbQuartzJobSupport.java

```
@@ -0,0 +1,113 @@
+  }
+
+  /**
+   * Given a ResourceList, find all the valid ids of referenced snapshots in that list
+   *
+   * @param resourceList the list in which to look for snapshotIds
+   * @return the list of unique ids in the provided list
+   */
+  protected List<UUID> extractSnapshotIds(ResourceList resourceList) {
+    return resourceList.getResources().stream()
+        .map(this::safeGetSnapshotId)
+        .filter(Objects::nonNull)
+        .distinct()
+        .toList();
+  }
+
+  /**
+   * Get the full list of all snapshot references for the current workspace. WSM returns these
+   * results paginated; this method retrieves pages from WSM and aggregates the results.
+   *
+   * @param pageSize number of results to return from WSM at once
+   * @return the full list of all snapshot references for the workspace.
+   */
+  protected ResourceList listAllSnapshots(int pageSize) {
+    final AtomicInteger offset = new AtomicInteger(0);
+    final int hardLimit =
+        10000; // under no circumstances return more than this many snapshots from WSM
+
+    ResourceList finalList = new ResourceList(); // collect our results
+
+    while (offset.get() < hardLimit) {
+      // get a page of results from WSM
+      RestClientRetry.RestCall<ResourceList> restCall =
+          (() -> wsmDao.enumerateDataRepoSnapshotReferences(workspaceId, offset.get(), pageSize));
+      ResourceList thisPage =
+          restClientRetry.withRetryAndErrorHandling(
+              restCall, "WSM.enumerateDataRepoSnapshotReferences");
+
+      // add this page of results to our collector
+      finalList.getResources().addAll(thisPage.getResources());
+
+      if (thisPage.getResources().size() < pageSize) {
+        // fewer results from WSM than we requested; this is the last page of results
+        return finalList;
+      } else {
+        // bump our offset and request another page of results
+        offset.addAndGet(pageSize);
+      }
+    }
+
+    throw new PfbImportException(
```

Is there any sensible way to fail faster if we'll exceed the hard limit?   Maybe by pushing the hard limit enforcement down to WSM?

Feels like this approach ends up doing a whole bunch of wasted work and would potentially fail slowly and cause resource waste/exhaustion when paired with some form of retries.

---

The only other approach I can think of is to pick a number (maybe not as high as 10000), and make a single request to WSM for that many resources - don't attempt to paginate. Do you think that would be better?

---

I think if by your guess this is a really really unlikely scenario that it's maybe not something to worry about and this might be the perfect safeguard...wait till it happens, then worry about it.

I don't think I understand well enough how these snapshots get written to WSM, but the only thing that I think might be better is if we could somehow head them off at write time (into WSM) rather than here at read time.